akka-stream/src/main/scala/akka/stream/scaladsl/SourceRef.scala

```diff
@@ -0,0 +1,272 @@
+        if (partnerRef.isDefined && i > 0) {
+          val addDemand = requestStrategy.requestDemand(receiveBuffer.used + localRemainingRequested)
+          // only if demand has increased we shoot it right away
+          // otherwise it's the same demand level, so it'd be triggered via redelivery anyway
+          if (addDemand > 0) {
+            localCumulativeDemand += addDemand
+            localRemainingRequested += addDemand
+            val demand = StreamRefs.CumulativeDemand(localCumulativeDemand)
+
+            log.debug("[{}] Demanding until [{}] (+{})", stageActorName, localCumulativeDemand, addDemand)
+            getPartnerRef ! demand
+            scheduleDemandRedelivery()
+          }
+        }
+      }
+
+      def scheduleDemandRedelivery() = scheduleOnce(DemandRedeliveryTimerKey, settings.demandRedeliveryInterval)
+
+      override protected def onTimer(timerKey: Any): Unit = timerKey match {
+        case SubscriptionTimeoutTimerKey ⇒
+          val ex = StreamRefs.StreamRefSubscriptionTimeoutException(
+            // we know the future has been competed by now, since it is in preStart
+            s"[$stageActorName] Remote side did not subscribe (materialize) handed out Sink reference [${promise.future.value}]," +
+              s"within subscription timeout: ${PrettyDuration.format(subscriptionTimeout.timeout)}!")
+
+          throw ex // this will also log the exception, unlike failStage; this should fail rarely, but would be good to have it "loud"
+
+        case DemandRedeliveryTimerKey ⇒
+          log.debug("[{}] Scheduled re-delivery of demand until [{}]", stageActorName, localCumulativeDemand)
+          getPartnerRef ! StreamRefs.CumulativeDemand(localCumulativeDemand)
+          scheduleDemandRedelivery()
+      }
+
+      lazy val initialReceive: ((ActorRef, Any)) ⇒ Unit = {
+        case (sender, msg @ StreamRefs.OnSubscribeHandshake(remoteRef)) ⇒
+          cancelTimer("SubscriptionTimeoutTimerKey")
+          observeAndValidateSender(remoteRef, "Illegal sender in SequencedOnNext")
+          log.debug("[{}] Received handshake {} from {}", stageActorName, msg, sender)
+
+          triggerCumulativeDemand()
+
+        case (sender, msg @ StreamRefs.SequencedOnNext(seqNr, payload)) ⇒
+          observeAndValidateSender(sender, "Illegal sender in SequencedOnNext")
+          observeAndValidateSequenceNr(seqNr, "Illegal sequence nr in SequencedOnNext")
+          log.debug("[{}] Received seq {} from {}", stageActorName, msg, sender)
+
+          tryPush(payload)
+          triggerCumulativeDemand()
+
+        case (sender, StreamRefs.RemoteStreamCompleted(seqNr)) ⇒
+          observeAndValidateSender(sender, "Illegal sender in RemoteSinkCompleted")
+          observeAndValidateSequenceNr(seqNr, "Illegal sequence nr in RemoteSinkCompleted")
+          log.debug("[{}] The remote stream has completed, completing as well...", stageActorName)
+
+          self.unwatch(sender)
+          completeStage()
```

Comment:
Isn't completion here skipping the buffer? Maybe it's the reason for the flaky test even?

Reply:
Good catch! Yes, it has to just flag here indeed